### PR TITLE
84 no results bug fix

### LINKED ISSUE #84: API helper no results error
# Describe the Bug
If no results are returned, there will be an error in extracting ids
# Additional Notes

# Story points to fix

*How much 'effort' this feature will take to implement. See the respective section in the [Main Dev Doc](https://github.com/cse110-fa21-group2/cse110-fa21-group2/wiki/Main-Developer-Doc#estimating-work-with-story-points) for more information*
1 point

### LINKED ISSUE #88: Check localStorage before fetching
# Is your feature request related to a problem? Please describe.
Unnecessary fetching when the recipes are already in Local Storage.
# Describe the solution you'd like 
Before fetching from API, check if the recipes are already in Local Storage.
# Describe alternatives you've considered

# Additional context

# Story Points
5pts

### PR DESCRIPTION
**Pull request by treeebooor**
## Details 
Fix no results error in getting recipes from API. Also added offset parameter to functions so you can have it return different results with the same search parameters. 

Also add checking local storage before fetching from API to prevent unnecessary fetches.
## New Changes
- apiHelpers.js
- testHelper.html

## Checklist
- [x ] I have requested 2+ code reviewers
- [ x] I have reviewed and commented my own code.
- [ x] My code follows the style guidelines of this project.
- [ x] I have added new tests as necessary.
- [ x] All new and existing tests pass successfully.

### CODE REVIEW

File: source/js/apiHelpers.js [hunk 1]
```diff
@@ -40,24 +40,30 @@ async function getDetailedRecipeInfoBulk(ids) {
...
+ *  (use random number so we dont get same results everytime)
  * @returns {Object} list of recipes with detailed info
  */
 // eslint-disable-next-line no-unused-vars
-async function getRecipesByName(query, num) {
+async function getRecipesByName(query, num, offset = 0) {
   return new Promise((resolve, reject) => {
     const queryFormatted = query.trim().replace(/\s+/g, '-').toLowerCase();
-    fetch(`https://${HOST}/recipes/complexSearch?&query=${queryFormatted}&number=${num}&sort=populatrity`, {
+    fetch(`https://${HOST}/recipes/complexSearch?&query=${queryFormatted}&number=${num}&sort=populatrity&offset=${offset}`, {
```
Comment (farisashai) [inline source/js/apiHelpers.js]: kinda unrelated but does this still work if parameters are missing, like if we just pass query and no number? might be good to put a default value there too but doesnt matter if its working. also there's a typo in popularity

File: source/js/apiHelpers.js [hunk 2]
```diff
@@ -40,24 +40,30 @@ async function getDetailedRecipeInfoBulk(ids) {
  */
 function extractIDs(data) {
   const { results } = data;
+  if (!results) {
+    return [];
+  }
   const ids = [];
   results.forEach((result) => {
+    // TODO: figure out how to check if current id is already in local storage
```
Comment (farisashai) [inline source/js/apiHelpers.js]: try checking `localStorage.getKey("recipeData')[id] == null` or undefined or something

File: source/js/apiHelpers.js [hunk 3]
```diff
@@ -155,11 +180,31 @@ async function getRecipesByCuisine(cuisine, num, sort) {
...
       })
       .catch((err) => {
-        console.log('Error in searching for recipes by cuisine.');
+        console.log('Error in searching for recipes by type.');
         reject(err);
       });
   });
 }
+
+// export functions?
```
Comment (farisashai) [inline source/js/apiHelpers.js]: yes pls export all of the functions. i think you can look at an example in some of the js files fred or misha wrote to learn the syntax.

File: source/js/apiHelpers.js (hunk 1 above)
Comment (treeebooor) [reply source/js/apiHelpers.js]: yea I can add default values

File: source/js/apiHelpers.js [hunk 4]
```diff
@@ -44,27 +42,28 @@ export async function getDetailedRecipeInfoBulk(idsToFetch) {
...
     if (recipeData[id] == undefined) {
       idsToFetch.push(id);
     } else {
-      idsInLocalStorage.push(id);
+      recipesInLocalStorage.push(recipeData[id]);
     }
   });
-  return [idsToFetch, idsInLocalStorage];
+  console.log(idsToFetch, recipesInLocalStorage);
+  return [idsToFetch, recipesInLocalStorage];
```
Comment (farisashai) [inline source/js/apiHelpers.js:62]: not completely sure how you're using this function so idk if this was intentional but i just wanted to emphasize that you're currently returning a list where both elements are lists. if you want to combine all of the elements instead, you can use the spread operator in js like this 
Screen Shot 2021-11-22 at 12 16 23 PM

Comment (treeebooor) [reply source/js/apiHelpers.js:62]: yea so Im trying to separate the ids that I need to fetch from API and the ids that are already found in local storage. I need them separate to call the API with idsToFetch and then recipesInLocalStorage will be the JSONs already found in local storage. 

Thats why I have two lists within a list. Is there a cleaner way to do this?

**Pull request closed without merging**